Kitodo-FileManagement/src/main/java/org/kitodo/filemanagement/locking/ImmutableReadFileManagement.java

```diff
@@ -0,0 +1,240 @@
+     * of numbers is inserted by the Java runtime and chosen so that it does not
+     * exist guaranteed.
+     *
+     * @param srcFile
+     *            the original file
+     * @return the derived file
+     * @throws IOException
+     *             if the file does not exist or if an error occurs in disk
+     *             access, e.g. because the write permission for the directory
+     *             is missing
+     */
+    private File deriveTempFile(File srcFile) throws IOException {
+        String prefix = srcFile.getName();
+        if (SystemUtils.IS_OS_WINDOWS) {
+            prefix = prefix.replace('.', '_');
+        }
+        prefix = prefix.concat("-");
+        String suffix = SystemUtils.IS_OS_WINDOWS ? ".tmp" : "~";
+        File directory = srcFile.getParentFile();
+        return File.createTempFile(prefix, suffix, directory);
+    }
+
+    /**
+     * Returns the immutable read copy for a user and a URI. If necessary, the
+     * file must be created at this point.
+     *
+     * @param user
+     *            user for whom the immutable read copy is to be returned
+     * @param uri
+     *            URI for which the immutable read copy is to be returned
+     * @return the immutable read copy
+     * @throws IOException
+     *             if the file does not exist or if an error occurs in disk
+     *             access, e.g. because the write permission for the directory
+     *             is missing
+     */
+    URI getImmutableReadCopy(String user, URI uri) throws IOException {
+        if (!urisGivenToUsers.containsKey(uri)) {
+            urisGivenToUsers.put(uri, new HashMap<>());
+        }
+        Map<String, Pair<URI, AtomicInteger>> userMapForURI = urisGivenToUsers.get(uri);
+        if (userMapForURI.containsKey(user)) {
+            Pair<URI, AtomicInteger> uriWithCount = userMapForURI.get(user);
+            uriWithCount.getValue().incrementAndGet();
+            return uriWithCount.getKey();
+        } else {
+            URI uriOfImmutableReadCopy = getUpToDateCopy(uri);
+            userMapForURI.put(user, Pair.of(uriOfImmutableReadCopy, new AtomicInteger(1)));
+            return uriOfImmutableReadCopy;
+        }
+    }
+
+    /**
+     * Returns the URI of a up-to-date read copy of the requested URI. Either,
+     * there is already one, or a copy process is triggered and the newly
+     * created copy is noted as up-to-date.
+     *
+     * @param originUri
+     *            URI for which the immutable read copy is to be returned
+     * @return the immutable read copy
+     * @throws IOException
+     *             if the file does not exist or if an error occurs in disk
+     *             access, e.g. because the write permission for the directory
+     *             is missing
+     */
+    private URI getUpToDateCopy(URI originUri) throws IOException {
+        URI upToDateCopy = upToDateCopies.get(originUri);
```

Comment:
Map.computeIfAbsent method could be used to replace the code like this:
      ```List<String> list = map.get(key);
            if (list == null) {
                list = new ArrayList<>();
                map.put(key, list);
             }```

Reply:
Changed 👍